Reverse cursor and iteration support

blambov opened the pull request.

michaeljmarshall commented on src/java/org/apache/cassandra/db/tries/Direction.java:
```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright DataStax, Inc.
```
```
08:28:51  /home/jenkins/workspace/ds-cassandra-pr-gate_PR-1079/.build/build-rat.xml:89: Some files have missing or incorrect license information. Check RAT report in /home/jenkins/workspace/ds-cassandra-pr-gate_PR-1079/build/rat.txt for more details! \n Unapproved licenses:
08:28:51  
08:28:51    /home/jenkins/workspace/ds-cassandra-pr-gate_PR-1079/src/java/org/apache/cassandra/db/tries/Direction.java
```

Looks like CI doesn't like this license.

blambov replied:
Done

michaeljmarshall reviewed (approved):
LGTM. We need to fix the license to make CI run. I did so in this PR https://github.com/datastax/cassandra/pull/1086 that targets the `vsearch` branch.

I reviewed this PR closely, but I am not familiar enough to critically review the low level details of this PR. One sign that it is correct: I just completed testing in https://github.com/datastax/cassandra/pull/1073, and the memtable reverse iteration works as expected.

blambov commented:
Apologies, pushed the license fix on the wrong branch (ds-tries-direction, if you are curious).

Added license, rebased and squashed, will commit after CI.

blambov commented:
Could you take another look? The two paths in `MemtableReadTrie` are not that different after all and could be combined so I did that to make my life easier for other coming changes.

michaeljmarshall commented on src/java/org/apache/cassandra/db/tries/SlicedTrie.java:
```diff
@@ -61,151 +61,279 @@ public SlicedTrie(Trie<T> source, ByteComparable left, boolean includeLeft, Byte
         this.includeRight = includeRight;
     }
 
+    static ByteSource add0(ByteSource src)
+    {
+        return new ByteSource()
+        {
+            boolean done = false;
+
+            @Override
+            public int next()
+            {
+                if (done)
+                    return END_OF_STREAM;
+                int next = src.next();
+                if (next != END_OF_STREAM)
+                    return next;
+                done = true;
+                return 0;
+            }
+        };
+    }
```
I've noticed this pattern elsewhere. Here is one such place:

https://github.com/datastax/cassandra/blob/f679bce038fabc2a0b69c74441a4b01121bd95b0/src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java#L271-L290

It's not a huge rush, but it seems like a generic class that could be added as a peer to `ByteSource.withTerminator`. That method uses `Multi`, which doesn't give us the same semantics as seen in this method and the one I reference above, so we would need to make it clear that we're simply adding a terminator.

michaeljmarshall reviewed (approved):
Looks good to me. Once we merge it, I'll cherry pick it to the `vsearch` branch.

michaeljmarshall commented:
Just bumped CI

michaeljmarshall commented:
vsearch cherry pick pr: https://github.com/datastax/cassandra/pull/1100
